File: water_pouring/water_pouring.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Fri Aug  3 09:02:36 2018

@author: tarik
"""
from collections import deque

class node:
    def __init__(self, quantities, parent):
        self.quantities = quantities
        self.parent = parent
        
    def __str__(self):
        return 'quantities={}, has_parent={}'.format(self.quantities, self.parent!=None)
    
def visited(node, quantities):
    while True:
        if node.quantities == quantities:
            return True
        elif node.parent == None:
            return False
        else:
            node = node.parent
        
from itertools import groupby
def get_quantity_by_count(quantities):
    return [(g[0], sum(1 for _ in g[1])) for g in groupby(sorted(quantities))]

def gen_solution_found(final_quantities):
    def solution_found(quantities):
        quantities_copy = quantities.copy()
        for quantity in final_quantities:
            if quantity in quantities_copy:
                quantities_copy.remove(quantity)
            else:
                return False
            
        return True
    
    return solution_found

def gen_solution(solution_node):
    solution_quantities_list=[]
    while True:
        solution_quantities_list.append(solution_node.quantities)
        solution_node = solution_node.parent
        if solution_node == None:
            break
        
    solution_quantities_list.reverse()
    return solution_quantities_list    
        
def solver(capacities, initial_quantities, final_quantities, allow_empty=False, allow_fill=False):
    solution_found = gen_solution_found(final_quantities)

    node_queue = deque([node(initial_quantities, None)])
    while True:
        if len(node_queue) == 0:
            return 
            
        current_node = node_queue.popleft()
        current_quantities = current_node.quantities
        for i, q_from in enumerate(current_quantities):
            for j, q_to in enumerate(current_quantities):
                capacity_to = capacities[j]
                if i != j and q_from != 0 and q_to < capacity_to:
                    new_quantities = current_quantities.copy()
                    q_transfer = min(capacity_to - q_to, q_from)
                    new_quantities[i]-=q_transfer
                    new_quantities[j]+=q_transfer
                    if not visited(current_node, new_quantities):
                        new_node = node(new_quantities, current_node)
                                                                       
                        if solution_found(new_quantities):
                            yield gen_solution(new_node)
                        else:
                            node_queue.append(new_node)


            if allow_empty and q_from > 0:
                new_quantities = current_quantities.copy()
                new_quantities[i] = 0
                if not visited(current_node, new_quantities):
                    new_node = node(new_quantities, current_node)
                                                                   
                    if solution_found(new_quantities):
                        yield gen_solution(new_node)
                    else:
                        node_queue.append(new_node)


            if allow_fill and q_from < capacities[i]:
                new_quantities = current_quantities.copy()
                new_quantities[i] = capacities[i]
                if not visited(current_node, new_quantities):
                    new_node = node(new_quantities, current_node)
                                                                   
                    if solution_found(new_quantities):
                        yield gen_solution(new_node)
                    else:
                        node_queue.append(new_node)


def solve(capacities, initial_quantities, final_quantities, allow_empty=False, allow_fill=False):
    solver_iterator = solver(capacities, initial_quantities, final_quantities, allow_empty=allow_empty, allow_fill=allow_fill)
    solution = next(solver_iterator, None)
    return solution

                       
def solve_and_print(capacities, initial_quantities, final_quantities, allow_empty=False, allow_fill=False, n_solutions=1):
    solver_iterator = solver(capacities, initial_quantities, final_quantities, allow_empty=allow_empty, allow_fill=allow_fill)
    i = 0
    while i<n_solutions:
        solution = next(solver_iterator, None)
        if solution != None:
            print('Solution found in {} moves:'.format(len(solution)-1))
            for move, quantities in enumerate(solution):
                print('{} - {}'.format(move, quantities))
            print(' ' * 20)
        else:
            if i==0:
                print('No solution found.')
            elif i+1<n_solutions:
                print('Only {} solutions found'.format(i+1))
            break
        i+=1

def solve_standard_problem(n_solutions=1):
    capacities=[8, 5, 3]
    initial_quantities=[8, 0, 0]
    final_quantities = [4, 4]
    solve_and_print(capacities, initial_quantities, final_quantities, n_solutions=n_solutions)
        

def find_toughest():
    capacities = [13, 7, 5]
    max_total_quantity = sum(capacities)
    max_steps = 0
    for i in range(min(capacities[0], max_total_quantity), -1, -1):
        for j in range(min(capacities[1], max_total_quantity-i), -1, -1):
            for k in range(min(capacities[2], max_total_quantity-i-j), -1, -1):
                initial_quantities = [i, j, k]
                total_quantity = sum(initial_quantities)
                for l in range(min(total_quantity, capacities[0])+1):
                    for m in range(min(total_quantity-l, capacities[1])+1):
                        n = total_quantity - l - m
                        if n > capacities[2]:
                            continue
                        final_quantities = [l, m, n]
                        solution = solve(capacities, initial_quantities, final_quantities)
                        if (solution != None and len(solution) > max_steps):
                            max_steps = len(solution)
                            print(f'{max_steps-1} steps')
                            print(capacities)
                            print(initial_quantities)
                            print(final_quantities)
                            print('-'*20)
                    
if __name__ == '__main__':
    #solve_standard_problem(n_solutions=3)
    find_toughest()
    """
    capacities = [13, 7, 5]
    initial_quantities = [13, 0, 0]
    final_quantities = [12, 0, 1]
    solve_and_print(capacities, initial_quantities, final_quantities, n_solutions=1)
    """
    

File: water_pouring/WaterPouring.jl
using DataStructures
using Query
using ProgressBars

mutable struct Node{T}
    quantities::Vector{T}
    parent::Union{Node{T}, Nothing}
# TODO: Implement show function
end

function visited(node::Node{T}, quantities::Vector{T}) :: Bool where T <: Integer
    while true
        if node.quantities == quantities
            return true
        elseif node.parent === nothing
            return false
        else
            node = node.parent
        end
    end
end

function gen_solution_found(final_quantities::Vector{T}) where T <: Integer
    groups = @groupby(_)(final_quantities)
    qs = [key(g) for g in groups]
    ns = [length(values(g)) for g in groups]
    function solution_found(quantities::Vector{T}) :: Bool
        all(i -> count(q -> q == qs[i], quantities) == ns[i], 1:length(qs))
    end
end

function gen_solution(solution_node::Node{T}) :: Vector{Vector{T}} where T
    solution_quantities_list=Vector{Vector{T}}(undef, 0)
    while true
        push!(solution_quantities_list, solution_node.quantities)
        solution_node = solution_node.parent
        if solution_node === nothing
            break
        end
    end

    reverse!(solution_quantities_list)
    return solution_quantities_list
end

function solver(capacities::Vector{T}, initial_quantities::Vector{T}, final_quantities::Vector{T}; allow_empty=false, allow_fill=false, n_solutions=1) where T <: Integer
    @debug "Capacities:\n$capacities"
    @debug "Initial Quantities:\n$initial_quantities"
    @debug "Final quantities:\n$final_quantities"
    solution_found = gen_solution_found(final_quantities)

    node_queue = Queue{Node{T}}()
    enqueue!(node_queue, Node{T}(initial_quantities, nothing))
    solutions = Vector{Vector{Vector{T}}}(undef, 0)
    while length(node_queue) > 0 && length(solutions) < n_solutions
        current_node = dequeue!(node_queue)
        current_quantities = current_node.quantities
        for (i, q_from) in enumerate(current_quantities)
            for (j, q_to) in enumerate(current_quantities)
                capacity_to = capacities[j]
                if i != j && q_from != 0 && q_to < capacity_to
                    new_quantities = copy(current_quantities)
                    q_transfer = min(capacity_to - q_to, q_from)
                    new_quantities[i]-=q_transfer
                    new_quantities[j]+=q_transfer
                    if !visited(current_node, new_quantities)
                        new_node = Node{T}(new_quantities, current_node)
                        if solution_found(new_quantities)
                            @debug "Solution found:\n$new_quantities"
                            push!(solutions, gen_solution(new_node))
                        else
                            @debug "New node with quantities:\n$new_quantities"
                            enqueue!(node_queue, new_node)
                        end
                    end
                end
            end

            if allow_empty && q_from > 0
                new_quantities = copy(current_quantities)
                new_quantities[i] = 0
                if !visited(current_node, new_quantities)
                    new_node = Node{T}(new_quantities, current_node)

                    if solution_found(new_quantities)
                        push!(solutions, gen_solution(new_node))
                    else
                        enqueue!(node_queue, new_node)
                    end
                end
            end

            if (allow_fill && q_from < capacities[i])
                new_quantities = current_quantities.copy()
                new_quantities[i] = capacities[i]
                if !visited(current_node, new_quantities)
                    new_node = Node{T}(new_quantities, current_node)

                    if solution_found(new_quantities)
                        push!(solutions, gen_solution(new_node))
                    else
                        enqueue!(node_queue, new_node)
                    end
                end
            end
        end
    end
    solutions
end

function solve(capacities::Vector{T}, initial_quantities::Vector{T}, final_quantities::Vector{T}, allow_empty=false, allow_fill=false) where T <: Integer
    solutions = solver(capacities, initial_quantities, final_quantities, allow_empty=allow_empty, allow_fill=allow_fill, n_solutions=1)
    if length(solutions) == 1
        return solutions[1]
    else
        return nothing
    end
end

function solve_and_print(capacities::Vector{T}, initial_quantities::Vector{T}, final_quantities::Vector{T}; allow_empty=false, allow_fill=false, n_solutions=1) where T <: Integer
    solutions = solver(capacities, initial_quantities, final_quantities, allow_empty=allow_empty, allow_fill=allow_fill)
    i = 0
    for solution in solutions
        if solution !== nothing
            println("Solution found in $(length(solution)-1) moves.")
            for (move, quantities) in enumerate(solution)
                println("$move - $quantities")
            end
            println(' ' ^ 20)
        else
            if i==0
                println("No solution found.")
            elseif i+1<n_solutions
                print("Only $(i+1) solutions found")
            end
            break
        end
        i+=1
        if i >= n_solutions
            break
        end
    end
end

function solve_standard_problem(n_solutions::Int)
    capacities=[8, 5, 3]
    initial_quantities=[8, 0, 0]
    final_quantities = [4, 4]
    solve_and_print(capacities, initial_quantities, final_quantities, n_solutions=n_solutions)
end

function find_toughest()
    capacities = [13, 7, 5]
    max_total_quantity = sum(capacities)-1
    max_steps = 0
    for i in ProgressBar(min(capacities[1], max_total_quantity):-1:0)
        for j in min(capacities[2], max_total_quantity-i):-1:0
            for k in min(capacities[3], max_total_quantity-i-j):-1:0
                initial_quantities = [i, j, k]
                total_quantity = sum(initial_quantities)
                for l in 0:min(capacities[1], max_total_quantity)
                    for m in 0:min(capacities[2], max_total_quantity-l)
                        n = total_quantity - l - m
                        if n > capacities[3]
                            continue
                        end
                        final_quantities = [l, m, n]
                        solution = solve(capacities, initial_quantities, final_quantities)
                        if (solution !== nothing && length(solution) > max_steps)
                            max_steps = length(solution)
                            println("$(max_steps-1) steps")
                            println(capacities)
                            println(initial_quantities)
                            println(final_quantities)
                            println('-'^20)
                        end
                    end
                end
            end
        end
    end
end

function main()
    solve_standard_problem(3)
    find_toughest()
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end
